### ISSUE TITLE
core::get_remaining_stream_length returns 32bit value

### ISSUE THREAD

**Issue by jhawkzz** [Azure/azure-storage-cpp#239]
First, want to say thanks for this awesome lib. It's super helpful!

It appears there's an issue with core::get_remaining_stream_length returning only a 32bit value for stream lengths.

When I attempt to upload either a cloud_file or cloud_block_blob, files larger than 4gb are truncated to 4gb. There's a workaround for cloud_file, but nothing for cloud_block_blob.

With cloud_file, I was able to fix this by manually providing the length of the stream (as read via GetFileSizeEx). In cloud_file.cpp, if a length is set, it just begins the upload (line 803).

With cloud_block_blob, unfortunately, if a length is provided there's an additional check using core::get_remaining_stream_length, and an exception is thrown because the 64bit value passed in via length is greater than the 32bit value obtained in core::get_remaining_stream_length(). (Lines 148 and then 160)

**Comment by katmsft** [comment]
Thanks for reporting the issue.
32/64bit compatibility is always messy in client libraries, not only the C++ one, since all the implicit conversions are not safe enough.
In this case, I wonder if there is a necessity of using 32bit platforms to deal with file larger than 4GB, the client library simply cannot guarantee all the dependencies behave correctly even if we make sure everything is fine in client library.
However, this issue is logged and we will discuss internally if we should scan for all the violations that would cause 32/64bit incompatibility. It would be very helpful for us to understand the importance of this, if you could provide more details of why this is a must for your scenario. Thanks again!

**Comment by jhawkzz** [comment]
Thanks for the response!

I agree that it would be fine to put documentation stating that if you intend on working with files over 4gb, you should compile for x64. (However, please note that there is still a bug when compiling for x64, which I note above in my original post.)

Our use case is for database backup. We have a SQL Server that generates a .bak file nightly, and we offload that to Azure Storage for redundancy. The file is ~90gb, so definitely goes past the 4gb limit! We decided to write our own utility because we have needs that tools like "AZCopy" don't offer out of the box.

Thanks again and please let me know if you need any further detail, or even the source code for our util to see how we're using the lib.

**Comment by katmsft** [comment]
Hi @EmmaZhu  is currently working on this, feel free to add any comment.
The method is actually return a 64bit value:
`utility::size64_t get_remaining_stream_length(concurrency::streams::istream stream)`
where
`typedef uint64_t size64_t;`
Why do you claim that this function returns 32bit value? Can you elaborate?
Also the code only returns `std::numeric_limits<utility::size64_t>::max()` if the stream is unseekable.

**Comment by jhawkzz** [comment]
@katmsft You are correct that it returns a size64_t value. However, the underlying function being called is _seekrdtoend_fsb in fileio_win32.cpp within the cpprest SDK. This function depends on Win32's SetFilePointer, which returns a LONG valuetype, which won't be greater than an int32. 

So in effect, a 32bit value is then being cast to 64bit.
